### PR TITLE
Get upstream changes

### PR DESCRIPTION
**Pull request by agitelzon**
Pull Request with merge conflicts fixed to get upstream changes.

### CODE REVIEW

File: mrq/version.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-VERSION = "0.9.10"
+VERSION = "0.9.29"
```
Comment (ghabrielv) [inline mrq/version.py]: Next version is 0.9.30

Comment (agitelzon) [reply mrq/version.py]: I haven't yet changed the version in this PR. That and over half of the items in the changes area are from commits that were applied upstream and now are getting re-applied on our repo. I'll post the actual changes.

**Comment by agitelzon** [comment]
Because a lot of commits are getting re-applied to our fork, the changes tab on this PR looks like there are more changes than there actually are. Here is a diff of the merged branch and the master branch:
```
❯ diff -Taur current_mrq_master merged_mrq --exclude .git
diff -Taur --exclude .git current_mrq_master/Makefile merged_mrq/Makefile
--- current_mrq_master/Makefile	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/Makefile	2022-07-29 09:59:42.000000000 -0500
@@ -56,7 +56,7 @@
		python mrq/dashboard/app.py

	stack:
-		mongod --smallfiles --noprealloc --nojournal &
+		mongod &
		redis-server &
		python mrq/dashboard/app.py &

diff -Taur --exclude .git current_mrq_master/docs/command-line.md merged_mrq/docs/command-line.md
--- current_mrq_master/docs/command-line.md	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/docs/command-line.md	2022-07-29 09:59:42.000000000 -0500
@@ -14,7 +14,7 @@
	 - `--trace_memory_output_dir`: Directory where to output .pngs with object graphs. Defaults to folder **memory_traces**.
	 - `--profile`: Run profiling on the whole worker. Defaults to **false**.
	 - `--mongodb_jobs, --mongodb`: MongoDB URI for the jobs, scheduled_jobs & workers database. Defaults to **mongodb://127.0.0.1:27017/mrq**.
-	 - `--mongodb_logs` :MongoDB URI for the logs database."0" will disable remote logs, "1" will use main MongoDB. Defaults to **1**
+	 - `--mongodb_logs` :MongoDB URI for the logs database."0" will disable remote logs, "1" will use main MongoDB. Defaults to **"1"**
	 - `--mongodb_logs_size`: If provided, sets the log collection to capped to that amount of bytes.
	 - `--redis`: Redis URI. Defaults to **redis://127.0.0.1:6379**.
	 - `--redis_prefix`: Redis key prefix. Defaults to "mrq".
@@ -46,7 +46,7 @@
	 - `--max_jobs`: Gevent:max number of jobs to do before quitting. Use as a workaround for memory leaks in your tasks. Defaults to **0**
	 - `--max_memory`: Max memory (in Mb) after which the process will be shut down. Use with `--processes [1-N]`
	                  to have the worker automatically respawn when this happens. Defaults to **1**
-	 - `--grenlets, --gevent, --g`: Max number of greenlets to use. Defaults to **1**.
+	 - `--greenlets, --gevent, --g`: Max number of greenlets to use. Defaults to **1**.
	 - `--processes, --p`: Number of processes to launch . Defaults to **0**.
	 - `--scheduler`: Run the scheduler. Defaults to **false**.
	 - `--scheduler_interval`: Seconds between scheduler checks. Defaults to **60** seconds, only ints are acceptable.
diff -Taur --exclude .git current_mrq_master/docs/configuration.md merged_mrq/docs/configuration.md
--- current_mrq_master/docs/configuration.md	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/docs/configuration.md	2022-07-29 09:59:42.000000000 -0500
@@ -18,7 +18,7 @@
	"""
	#MongoDB settings
	MONGODB_JOBS = "mongodb://127.0.0.1:27017/mrq" # MongoDB URI for the jobs, scheduled_jobs & workers database.Defaults to mongodb://127.0.0.1:27017/mrq
-	MONGODB_LOGS = 1 #MongoDB URI for the logs database."0" will disable remote logs, "1" will use main MongoDB.Defaults to 1
+	MONGODB_LOGS = "1" #MongoDB URI for the logs database."0" will disable remote logs, "1" will use main MongoDB.Defaults to 1
	MONGODB_LOGS_SIZE = None #If provided, sets the log collection to capped to that amount of bytes.

	#Redis settings
@@ -57,7 +57,7 @@
	MAX_JOBS = 0 #Gevent:max number of jobs to do before quitting. Workaround for memory leaks in your tasks. Defaults to 0
	MAX_TIME = 0 # max number of seconds a worker runs before quitting
	MAX_MEMORY = 1 #Max memory (in Mb) after which the process will be shut down. Use with PROCESS = [1-N] to have the worker automatically respawned when this happens.Defaults to 1
-	GRENLETS = 1 #Max number of greenlets to use.Defaults to 1.
+	GREENLETS = 1 #Max number of greenlets to use.Defaults to 1.
	PROCESSES = 0 #Number of processes to launch.Defaults to 0.
	SCHEDULER = False #Run the scheduler.Defaults to False.
	SCHEDULER_INTERVAL = 60 #Seconds between scheduler checks.Defaults to 60 seconds, only ints are acceptable.
diff -Taur --exclude .git current_mrq_master/docs/queue-performance.md merged_mrq/docs/queue-performance.md
--- current_mrq_master/docs/queue-performance.md	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/docs/queue-performance.md	2022-07-29 09:59:42.000000000 -0500
@@ -174,7 +174,7 @@

	We are now close to 9000 jobs per second, maxing-out the local CPUs again!

-	From there on, the sky is the limit! You should be able to run thousands of workers accross hundreds of machines before maxing-out a high-performance Redis instance.
+	From there on, the sky is the limit! You should be able to run thousands of workers across hundreds of machines before maxing-out a high-performance Redis instance.

	Beyond that, using using multiple queues on a [Redis Cluster](https://redis.io/topics/cluster-tutorial) will definitely allow you to run several million jobs per second. If you do, please drop us a line ;-)

diff -Taur --exclude .git current_mrq_master/mrq/agent.py merged_mrq/mrq/agent.py
--- current_mrq_master/mrq/agent.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/mrq/agent.py	2022-07-29 09:59:42.000000000 -0500
@@ -8,7 +8,13 @@
	import traceback
	from collections import defaultdict
	from bson import ObjectId
-	from redis.lock import LuaLock
+
+	try:
+	    from redis.lock import LuaLock
+	except ImportError:
+	    # Change name to avoid NameError raised when use of LuaLock at line 147
+	    from redis.lock import Lock as LuaLock
+
	from .processes import Process, ProcessPool
	from .utils import MovingETA, normalize_command
	from .queue import Queue
diff -Taur --exclude .git current_mrq_master/mrq/basetasks/utils.py merged_mrq/mrq/basetasks/utils.py
--- current_mrq_master/mrq/basetasks/utils.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/mrq/basetasks/utils.py	2022-07-29 09:59:42.000000000 -0500
@@ -52,6 +52,8 @@
	            if self.params.get(k):
	                if isinstance(self.params[k], (list, tuple)):
	                    query[k] = {"$in": list(self.params[k])}
+	                elif k == "queue" and self.params[k][-1] == "/":
+	                    query[k] = {"$regex": "^%s" % self.params[k]}
	                else:
	                    query[k] = self.params[k]
	            if query.get("worker"):
@@ -117,21 +119,20 @@
	            # In this case we could also loose some jobs that were queued after
	            # the MongoDB update. They will be "lost" and requeued later like the other case
	            # after the Redis BLPOP
-	            if list(query.keys()) == ["queue"]:
+	            if list(query.keys()) == ["queue"] and isinstance(query["queue"], basestring):
	                Queue(query["queue"]).empty()

	        elif action in ("requeue", "requeue_retry"):

	            # Requeue task by groups of maximum 1k items (if all in the same
	            # queue)
-	            cursor = self.collection.find(query, projection=["_id", "queue"])
+	            status_query = query.get("status")
+	            if not status_query:
+	                query["status"] = {"$ne": "queued"}

-	            # We must freeze the list because queries below would change it.
-	            # This could not fit in memory, research adding {"stats": {"$ne":
-	            # "queued"}} in the query
-	            fetched_jobs = list(cursor)
+	            cursor = self.collection.find(query, projection=["_id", "queue"])

-	            for jobs in group_iter(fetched_jobs, n=1000):
+	            for jobs in group_iter(cursor, n=1000):

	                jobs_by_queue = defaultdict(list)
	                for job in jobs:
diff -Taur --exclude .git current_mrq_master/mrq/processes.py merged_mrq/mrq/processes.py
--- current_mrq_master/mrq/processes.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/mrq/processes.py	2022-07-29 09:59:42.000000000 -0500
@@ -37,10 +37,10 @@
	                self.shutdown_graceful()

	        # First time CTRL-C, try to shutdown gracefully
-	        gevent.signal(signal.SIGINT, request_shutdown_graceful)
+	        gevent.signal_handler(signal.SIGINT, request_shutdown_graceful)

	        # User (or Heroku) requests a stop now, just mark tasks as interrupted.
-	        gevent.signal(signal.SIGTERM, request_shutdown_now)
+	        gevent.signal_handler(signal.SIGTERM, request_shutdown_now)


	class ProcessPool(object):
diff -Taur --exclude .git current_mrq_master/mrq/worker.py merged_mrq/mrq/worker.py
--- current_mrq_master/mrq/worker.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/mrq/worker.py	2022-07-29 09:59:42.000000000 -0500
@@ -15,7 +15,13 @@
	import json as json_stdlib
	import ujson as json
	from bson import ObjectId
-	from redis.lock import LuaLock
+
+	try:
+	    from redis.lock import LuaLock
+	except ImportError:
+	    # Change name to avoid NameError raised when use of LuaLock at line 151
+	    from redis.lock import Lock as LuaLock
+
	from collections import defaultdict
	from mrq.utils import load_class_by_path

@@ -222,13 +228,21 @@
	          time.sleep(self.config["paused_queues_refresh_interval"])

	    def get_memory(self):
-	        mmaps = self.process.memory_maps()
-	        mem = {
-	            "rss": sum([x.rss for x in mmaps]),
-	            "swap": sum([getattr(x, 'swap', getattr(x, 'swapped', 0)) for x in mmaps])
-	        }
-	        mem["total"] = mem["rss"] + mem["swap"]
-	        return mem
+
+	        try:
+
+	            mmaps = self.process.memory_maps()
+	            mem = {
+	                "rss": sum([x.rss for x in mmaps]),
+	                "swap": sum([getattr(x, 'swap', getattr(x, 'swapped', 0)) for x in mmaps])
+	            }
+	            mem["total"] = mem["rss"] + mem["swap"]
+	            return mem
+
+	        # memory_maps is unavailable on macOS
+	        # https://github.com/pricingassistant/mrq/issues/228
+	        except Exception as e:
+	            return {"total": 0, "rss": 0, "swap": 0}

	    def get_worker_report(self, with_memory=False):
	        """ Returns a dict containing all the data we can about the current status of the worker and
@@ -342,7 +356,7 @@

	        if self.config["report_file"]:
	            with open(self.config["report_file"], "wb") as f:
-	                f.write(bytes(json.dumps(report, ensure_ascii=False), 'utf-8'))  # pylint: disable=no-member
+	                f.write(bytes(json.dumps(report, ensure_ascii=False, default=str), 'utf-8'))  # pylint: disable=no-member

	        if "_id" in report:
	            del report["_id"]
@@ -437,7 +451,7 @@
	                # We might be dequeueing a new subqueue. Double check that we don't have anything more to do
	                outcome, dequeue_jobs = self.work_once(free_pool_slots=1, max_jobs=None)

-	                if outcome is "wait" and dequeue_jobs == 0:
+	                if outcome == "wait" and dequeue_jobs == 0:
	                    break

	    def work(self):
diff -Taur --exclude .git current_mrq_master/setup.py merged_mrq/setup.py
--- current_mrq_master/setup.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/setup.py	2022-07-29 09:59:42.000000000 -0500
@@ -61,7 +61,6 @@
	    classifiers=[
	        "Programming Language :: Python",
	        "Programming Language :: Python :: 2.7",
-	        "Programming Language :: Python :: 3.4",
	        "Programming Language :: Python :: 3.5",
	        "Programming Language :: Python :: 3.6",
	        "Programming Language :: Python :: 3.7",
diff -Taur --exclude .git current_mrq_master/tests/conftest.py merged_mrq/tests/conftest.py
--- current_mrq_master/tests/conftest.py	2022-07-29 09:56:10.000000000 -0500
+++ merged_mrq/tests/conftest.py	2022-07-29 09:59:42.000000000 -0500
@@ -373,7 +373,7 @@

	@pytest.fixture(scope="function")
	def mongodb(request):
-	    cmd = "mongod --smallfiles --noprealloc --nojournal"
+	    cmd = "mongod --nojournal"
	    if os.environ.get("STACK_STARTED"):
	        cmd = "sleep 1h"
	    return MongoFixture(request, cmd, wait_port=27017, quiet=True)
@@ -381,7 +381,7 @@

	@pytest.fixture(scope="function")
	def mongodb_with_journal(request):
-	    cmd = "mongod --smallfiles --noprealloc"
+	    cmd = "mongod"
	    if os.environ.get("STACK_STARTED"):
	        cmd = "sleep 1h"
	    return MongoFixture(request, cmd, wait_port=27017, quiet=True)
```

**Comment by orlandobcrra** [comment]
@ghabrielv @agitelzon I [made some changes to MRQ](https://github.com/apploitech/mrq/pull/9/files) (new version) that are not being used yet on our system (our app still using the old version of). 
those changes are not going to affect how MRQ works currently and will not break any task nor worker server, but those will break unit tests of python_api and integrations. I have the changes on a PR to fix those tests for python API, and have to do it for integrations. if this bump is urgent, let me know.

**Review by orlandobcrra — COMMENTED** [review]
few questions: 
1. what benefits bring those changes? 
2. do we need those changes?
3. since we are moving away from MRQ, are there new features we need to consider for the new serverless approach?

**Comment by agitelzon** [comment]
> few questions:
> 
> 1. what benefits bring those changes?
> 2. do we need those changes?
> 3. since we are moving away from MRQ, are there new features we need to consider for the new serverless approach?

My answers:
1. These changes are to put us inline with the upstream project. At minimum, I saw a fix for exceptions being thrown with LuaLock and a try/catch block added
2. We don't need these changes, but I was hoping maybe they would fix one of the deadlocks we experience
3. Overall, we need better exception handling and logging when things go wrong

**Pull request closed without merging**